### PR TITLE
Mentions

### LINKED ISSUE #92: Mentions
We need to store unread @mentions on the backend so we can display unread ones to the user. Like Discord. I'd like to label this as vital as honestly it's a feature I _really_ miss using Matrix.

 It's also worth keeping in mind that the way Matrix/Riot does this is the same as IRC, which is just to gather a list of mentions **client-side** and highlight them differently. This means you can't see mentions from the history of the room you're in, or globally :(

---

Discord implements this by formatting messages specially before they are sent by the client. It will convert `@alex` to `@<1234567890>` (or whatever _alex_'s user ID is) and then send _that_ to the server. The server will parse `@<[0-9]+>`s present in the message text and add that mention to the database. It'll also ping the user.

This is cool, because it ensures that even if you change your username - which you can't at the moment! - your mentions will stay intact. Oh, and we can't forget `@everyone`!

This can go along with a "mentions only" channel notifications feature (another thing to add!!), as opposed to just 'mute' or 'all.'

### LINKED ISSUE #227: Sidebars should be scrollable
Who wants to CSS position-sticky headers like "users" and "channels", which is now [supported on everything](https://developer.mozilla.org/en-US/docs/Web/CSS/position#Browser_compatibility)? Or maybe that's a dumb idea :P

Blocked by #226, but could be implemented into that (ping @heyitsmeuralex, feel free to unassign yourself if you don't want to do this in that PR).

### PR DESCRIPTION
**Pull request by heyitsmeuralex**
<sup>"mention" doesn't look like a word anymore <sup>help</sup></sup>

Adds mentions; fixes #92. Includes respective APIs and events (similar to the pins-related ones). Mentions are formatted as `<@user_id>` and are tracked by the server :package: 

Also:
* fixes #227 
* updated [mrk](/heyitsmeuralex/mrk) to `2.0.0` - it's now factory-function based and supports a `require('mrk.js/async')` Promise mode. We currently can't do async rendering so it's not very useful.

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -713,8 +716,10 @@ Model:
 
   "online": boolean,
 
+  "mentions": [ Message ] // List of messages that mention this user
```
Comment (towerofnix) [inline docs/api.md]: Missing comma.

File: packages/server/api.js [hunk 2]
```diff
@@ -50,6 +51,56 @@ module.exports = async function attachAPI(app, {wss, db, dbDir}) {
     }
   }
 
+  const handleMentionsInMessage = async function (message) {
+    for (let userID of new Set(await getMentionsFromMessageContent(message.text))) {
```
Comment (towerofnix) [inline packages/server/api.js]: Should be `for (const userID`, not `let`.

File: packages/server/api.js [hunk 3]
```diff
@@ -50,6 +51,56 @@ module.exports = async function attachAPI(app, {wss, db, dbDir}) {
...
+      }
+    }
+  }
+
+  const handleUnmentionsInMessage = async function (message, newText) {
+    const mentionedOld = await getMentionsFromMessageContent(message.text)
+    const mentionedNew = await getMentionsFromMessageContent(newText)
+    const unmentionedInNew = mentionedOld.filter(id => !mentionedNew.includes(id))
+
+    for (let userID of unmentionedInNew) {
```
Comment (towerofnix) [inline packages/server/api.js]: Again, should be `for (const of`.

File: packages/server/api.js [hunk 4]
```diff
@@ -467,6 +520,8 @@ module.exports = async function attachAPI(app, {wss, db, dbDir}) {
         return
       }
 
+      await handleUnmentionsInMessage(oldMessage, text)
```
Comment (towerofnix) [inline packages/server/api.js]: Does this remove old mentions from the message's database document? (For example if I edit from `<@a> <@b>` to `<@a>`, does `b` get removed from the document?)

File: docs/api.md (hunk 1 above)
Comment (heyitsmeuralex) [reply docs/api.md]: We don't have commas in [some?] other models of the API docs; plus it's not particularly standard to anyone but us. I'd like to leave it off if I'm honest.

File: packages/server/api.js (hunk 4 above)
Comment (heyitsmeuralex) [reply packages/server/api.js]: As far as I know, yes.

File: docs/api.md (hunk 1 above)
Comment (heyitsmeuralex) [reply docs/api.md]: Just committed to make everything more consistent.

Comment (towerofnix) [reply docs/api.md]: The problem here is that there are more properties after this one (`authorized`, `email`).

File: packages/client/src/components/sidebar-right.js [hunk 5]
```diff
@@ -175,6 +175,13 @@ const component = (state, emit) => {
               <div
                 class='Sidebar-list-item UserList-user ${user.online ? 'is-online' : 'is-offline'}'
                 title='${user.username}${user.flair ? ` {${user.flair}}` : ''} (${user.online ? 'Online' : 'Offline'})'
+                onclick=${() => {
+                  const textarea = document.querySelector('.MessageEditor-textarea')
+
+                  if (textarea) {
+                    textarea.value += '@' + user.username
```
Comment (towerofnix) [inline packages/client/src/components/sidebar-right.js]: This should insert a space before the `@` if (the textarea is not empty && the last character is not a space / at the start of the line).

Also, this should totally insert at wherever the cursor is positioned in the textarea, but don't worry about that now!! It's super a pain to implement.

File: packages/client/src/components/sidebar-right.js [hunk 6]
```diff
@@ -179,7 +179,12 @@ const component = (state, emit) => {
                   const textarea = document.querySelector('.MessageEditor-textarea')
 
                   if (textarea) {
-                    textarea.value += '@' + user.username
+                    if (textarea.value.length > 0) {
+                      const lastChar = textarea.value[textarea.value.length - 1]
+                      if (lastChar !== ' ') textarea.value += ' '
```
Comment (towerofnix) [inline packages/client/src/components/sidebar-right.js]: Should be `lastChar !== ' ' && lastChar !== '\n'`.

**Pull request merged**